reference/reference.go

```diff
@@ -0,0 +1,181 @@
+package reference
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/docker/distribution/digest"
+	distreference "github.com/docker/distribution/reference"
+	"github.com/docker/docker/image/v1"
+)
+
+const (
+	// DefaultTag defines the default tag used when performing images related actions and no tag string is specified
+	DefaultTag = "latest"
+	// DefaultHostname is the default built-in hostname
+	DefaultHostname = "docker.io"
+	// LegacyDefaultHostname is automatically converted to DefaultHostname
+	LegacyDefaultHostname = "index.docker.io"
+	// DefaultRepoPrefix is the prefix used for defautl repositories in default host
+	DefaultRepoPrefix = "library/"
+)
+
+// Reference is an opaque object reference identifier that may include
+// modifiers such as a hostname, name, tag, and digest.
+type Reference interface {
+	// String returns the full reference
+	String() string
+}
+
+// Named is an object with a full name
+type Named interface {
+	Reference
+	Name() string
+}
```

Comment:
I'm not sure if this is a good idea or not, but I wonder if we should add an extra method to `Named` so it is not directly interchangeable with `distribution/reference`'s `Named`.

The original purpose of the reference types was to use the type system to make sure that references are valid instances of their specific types. But now it's possible to accidentally pass an unvalidated, unnormalized `Named` reference into a function that expects the validation and normalization simply by importing the wrong `reference` package and calling `ParseNamed`.

If we did this, this package would need to wrap `Named` objects returned by distribution in a struct which has that extra method defined. It would add a bit of code to this file, but make for safer code elsewhere.